### PR TITLE
--skip-action-mailer option doesn't work for "rails plugin new"

### PR DESCRIPTION
**Pull request by piecehealth**
### Summary

Fix `--skip-action-mailer` option doesn't work for `rails plugin new`.

#### Update
Also fix `--skip-active-job` option doesn't work for `rails new` and `rails plugin new`.

### CODE REVIEW

File: railties/test/generators/plugin_generator_test.rb [hunk 1]
```diff
@@ -226,6 +226,21 @@ def test_skip_javascript
     end
   end
 
+  def test_skip_action_mailer_with_mountable
+    run_generator [destination_root, "--mountable", "--skip-action-mailer"]
+    assert_no_directory "app/mailers"
```
Comment (tgxworld) [inline railties/test/generators/plugin_generator_test.rb:231]: Hmm we're already testing that the `app/mailers` directory does not exists in 

https://github.com/rails/rails/blob/bcfa51fba6ee529fbe357a183fd5bc8bebf13513/railties/test/generators/shared_generator_tests.rb#L304-L305

Any ideas why that test is passing?

Comment (piecehealth) [reply railties/test/generators/plugin_generator_test.rb:231]: The test specified neither `--full` option nor `--mountable` option, so the `app` folder didn't create at all.

Comment (tgxworld) [reply railties/test/generators/plugin_generator_test.rb:231]: Got it :) Do we have to ensure that the `dummy/app/mailers` directory is not created too?

Comment (tgxworld) [reply railties/test/generators/plugin_generator_test.rb:231]: Oops sorry I was testing without `--skip-action-mailer`. Looks like it does remove the `test/dummy/app/mailers` directory.

**Comment by tgxworld** [comment]
@piecehealth This is out of scope based on the PR's title but I see that a `jobs` folder is created by default too. We should probably  apply a similar patch when `--skip-active-job` is passed.

**Comment by piecehealth** [comment]
@tgxworld it seems `rails new project_name` doesn't provide `--skip-active-job` either, maybe should take another PR to fix that.

**Comment by tgxworld** [comment]
I think the option doesn't exist in Rails 6.0 but I do see it on latest

https://github.com/rails/rails/blob/bcfa51fba6ee529fbe357a183fd5bc8bebf13513/railties/lib/rails/generators/app_base.rb#L55-L56

**Comment by piecehealth** [comment]
🤔  `rails new --help` doesn't show the `skip-active-job` option `rails new project --skip-active-job` doesn't remove the folder. I will see whether it's easy to fix and combine two fixes together when I get time.

**Comment by tgxworld** [comment]
> I will see whether it's easy to fix and combine two fixes together when I get time.

No problem. Let me know if you can get to it otherwise I can take it as well.

**Comment by piecehealth** [comment]
@tgxworld `--skip-active-job` is a newly added by https://github.com/rails/rails/commit/7a321d8429e3d7b9d8aa97a44ce4a4b6cb62de70 so I didn't say the option on my local 😓 

Anyway, just pushed another commit to fix it.

File: railties/test/generators/plugin_generator_test.rb [hunk 2]
```diff
@@ -226,6 +226,37 @@ def test_skip_javascript
...
+
+  def test_skip_active_job_with_api_and_mountable
+    run_generator [destination_root, "--api", "--mountable", "--skip-active-job"]
+    assert_no_directory "app/jobs"
+  end
+
+  def test_skip_active_job_with_full
+    run_generator [destination_root, "--full", "--skip-active-job"]
+    assert_no_directory "app/jobs"
+  end
```
Comment (tgxworld) [inline railties/test/generators/plugin_generator_test.rb:257]: I was wondering if it'll be better to just combine the tests for `--skip-active-job` and `--skip-action-mailer` together. No need to run the generator more times than necessary.

Comment (piecehealth) [reply railties/test/generators/plugin_generator_test.rb]: make sense 👍

File: railties/lib/rails/generators/rails/plugin/plugin_generator.rb [hunk 3]
```diff
@@ -19,15 +19,29 @@ def rakefile
...
-          directory "app"
+          excluded_directories = []
+          excluded_directories << "mailers" if options[:skip_action_mailer]
+          excluded_directories << "jobs" if options[:skip_active_job]
+
+          if excluded_directories.empty?
+            directory "app"
+          else
+            directory "app", exclude_pattern: %r{app/(#{excluded_directories.join("|")})}
+          end
```
Comment (eugeneius) [inline railties/lib/rails/generators/rails/plugin/plugin_generator.rb:36]: Using `exclude_pattern` here seems to be causing a lot of complexity, and dynamically generating a regex makes the logic quite hard to follow. What if we did the same thing as the app generator and [removed](https://github.com/rails/rails/blob/b39ea7cc3662345ad8efbe496db5ae594449f85f/railties/lib/rails/generators/rails/app/app_generator.rb#L504) the [directories](https://github.com/rails/rails/blob/b39ea7cc3662345ad8efbe496db5ae594449f85f/railties/lib/rails/generators/rails/app/app_generator.rb#L496) after copying them instead? The diff would be much simpler:

```patch
diff --git a/railties/lib/rails/generators/rails/plugin/plugin_generator.rb b/railties/lib/rails/generators/rails/plugin/plugin_generator.rb
index d59296a64e..7db6d9b63f 100644
--- a/railties/lib/rails/generators/rails/plugin/plugin_generator.rb
+++ b/railties/lib/rails/generators/rails/plugin/plugin_generator.rb
@@ -25,10 +25,14 @@ def app
           directory "app"
           empty_directory_with_keep_file "app/assets/images/#{namespaced_name}"
         end
+
+        remove_dir "app/mailers" if options[:skip_action_mailer]
+        remove_dir "app/jobs" if options[:skip_active_job]
       elsif full?
         empty_directory_with_keep_file "app/models"
         empty_directory_with_keep_file "app/controllers"
-        empty_directory_with_keep_file "app/mailers"
+        empty_directory_with_keep_file "app/mailers" unless options[:skip_action_mailer]
+        empty_directory_with_keep_file "app/jobs" unless options[:skip_active_job]

         unless api?
           empty_directory_with_keep_file "app/assets/images/#{namespaced_name}"
```

Comment (piecehealth) [reply railties/lib/rails/generators/rails/plugin/plugin_generator.rb]: @eugeneius PR updated, please take another look, thanks.

**Comment by eugeneius** [comment]
Thanks!

**Comment by udaykadaboina** [comment]
Just curious as to which release is this code available? I have 6.0.3.3 and doesn't seem to have skip-active-job option.

**Comment by tgxworld** [comment]
@udaykadaboina  This code will only be available in the 6.1 release.
